### ISSUE TITLE
Darktable corrupting RAW files Mac OS 10.13.6 

### ISSUE THREAD

**Issue by swsounds** [darktable-org/darktable#7173]
**Describe the bug**
I have been trying to use DT 3.2.1 on a Mac OSX 10.13.6. for about a week.  I have been using it on W10 for 6 months or so with very few issues (maybe 6 crashes in total and no corrupted RAWS). I have had a number of crashes usually when I am trying to zoom in on an image in DR.  I have also had a few issues after crashes where images become corrupted and I have got a DamagedFiles Folder with 4 RAWs listed in it. However, this morning I tried to load an image to DR and it just came up with the error message shown in the screen grab (on one occasion it mentioned White balance but that has not reappeared. It automatically goes back to LT and the skull appears. I have trashed the XMP and reimported the RAW but nothing changes. 

I am not a developer but I downloaded EXIFTOOL as suggested and have tried running it but I have no idea what is going on with it so I cannot review the file. However, I have loaded the same image into W10 DT 3.2.1 and it is fine. 

**To Reproduce**
Provide detailed steps that can reproduce the behavior, such as:
1. Go to '.LT'
2. Click on image to open

3. See errors shown in image

**Expected behavior**
Image should open in DR

**Screenshots**

![image](https://user-images.githubusercontent.com/75428372/101051076-fb1b0180-357c-11eb-957d-e3f4556d8000.png)

**Platform (please complete the following information):**
 - Darktable Version: -- 3.2.1 
 - OS:  OSX 10.13.6
 -  OpenCL no? 
 - Intel HD Graphics 4000 1536 MB graphics
Mac Min i7 Quad 2012 8GB RAM 

**Additional context**
I have DT on a W10 laptop and this image is perfectly happy on there.
This has happened to 3 other images, usually after a crash, this one no crash that I was aware of:
I have got the RAW from the PC and loaded into a new folder and imported and applied the style and it is perfectly happy .
I did try trashing the XMP from the original image and reimporting but it gave me the same error and will not load.
I have installed DT on my Mac from the standard downloader.
The other issue I have been having over the last few days using it on the Mac is that in LT view I will get a few thumbnails that just won't display properly, they will be all black or partially black or partially scrambled yet when I load them they will be fine and when LT refreshes it all looks fine. 
I didn't think DT could change the RAW file but it appears that it is corrupting it. I have only been using DT on here for a weeks worth of processing, although I did load the program on here at the same time I started on the PC. 
I have upgraded to OS10.13 from 10.12 in the last month. 
I have had more problems on this machine in one week than in 6 months on the laptop but I would prefer to use the Mac because of the larger screens. 
The only thing on here that is not pure Mac is the XP-Pen tablet - a Deco 01v2. Which I am not sure is completely happy. I don't think it is causing the crashes though...
 
Is there anything obvious that I am missing? 
I hope I have provided sufficient information. 

Thanks

Steve

**Comment by ralfbrown** [comment]
If a copy of the RAW file works fine on another installation, you probably wound  up with some corrupted data in the dt database, e.g. a broken xmp sidecar got loaded.

To check this possibility, you can add "--library :memory:" to the dt commandline, and it will run with an in-memory database -- this starts empty, and goes away when you exit dt.  Try importing the image and check for problems.

If you do any edits while running with the in-memory database, the xmp sidecar will still be updated even though the database itself goes away.  But if there are no problems, that updated xmp can be the basis for fixing your main database -- ensure that "check for updated sidecar files" is enabled in preferences, then start dt without overriding the library.  It should find the new xmp and offer to copy it into the database.

**Comment by swsounds** [comment]
Hi Ralf

I thought this too but I trashed the initial XMP and reimported the raw that was there to create a new XMP and this came up with the same error message. 
After this I trashed both of them and got the copy of the RAW from my PC and imported it and it worked. This leads me to believe that the RAW was damaged not the XMP.

How do I add to the dt command line? 
I have found where to look for updated amp on start up but how do I start dt without overriding the library?

Thanks again

Steve

**Comment by parafin** [comment]
Raw file may have been damaged, but it's very unlikely that darktable is the cause.

**Comment by swsounds** [comment]
Hi Parafin

I agree that RAW damaged but how if not DT? 

All I have done is put file onto Hard Drive from SD and then imported it and applied a Style. 

I have had at least 4 corrupted RAW files this week on the Mac and not a single one on the PC in 6 months and DT has been crashing on the Mac.

I look forward to hearing further.

Thanks

Steve

**Comment by parafin** [comment]
darktable doesn't write to raw files ever.
Files can get damaged by other software or due to hardware failure.

**Comment by swsounds** [comment]
Which is what I thought but might a dt crash corrupt an open file?
On the basis that I haven't opened these files with any other software what sort of hardware failure would cause this?
Thanks

**Comment by chhil** [comment]
Can you do a binary diff on the raw files to determine if anything changed?

**Comment by ptilopteri** [comment]
dt does NOT write to raw files in any case, even when crashing.

**Comment by ralfbrown** [comment]
> How do I add to the dt command line?

In a Terminal window, instead of typing just "darktable", type "darktable --library :memory:".

> I have found where to look for updated amp on start up but how do I start dt without overriding the library?

The --library flag only affects the current run, so just start darktable normally.

**Comment by MStraeten** [comment]
can you provide the file? So someone can check if it’s really defect on osx ...

**Comment by swsounds** [comment]
Ralf,

Thanks for that information. I am not too hot on command line so that is very helpful. 

When I have run the memory in cmd line how do I check it?

Thanks again.

Steve

**Comment by swsounds** [comment]
chhil and MStraeten,

Thanks for your suggestions. I don't know how to run a binary diff - sorry - I know what you are suggesting in theory but have no idea where to start in practice.

Here is the original CR2 file that was apparently corrupted.  I have zipped it because CR2 files are not supported for the upload

The image is not important or that special. It just doesn't work and I'd like to know why...

[IMG_2146.CR2.zip](https://github.com/darktable-org/darktable/files/5644008/IMG_2146.CR2.zip)

I look forward to hearing what anyone thinks about this.

Thanks

**Comment by parafin** [comment]
Can you also provide the non-corrupted file so that we could compare it?

**Comment by MStraeten** [comment]
on osx:`exiv2 -pv ~/Downloads/IMG_2146.CR2`
Exiv2 exception in print action for file xxx/Downloads/IMG_2146.CR2:
xxx/Downloads/IMG_2146.CR2: The file contains data of an unknown image type

darktable can’t open this cr2 file since it isn't identified as a known image type

**Comment by parafin** [comment]
First 7 MB of the file is just zeros. I would say it's a filesystem corruption or disk failure. Or some software bug somewhere of course, but as I said it's highly unlikely to be in darktable since it never writes anything to the original files.

**Comment by TurboGit** [comment]
As @parafin said we never write to RAW files, I haven't checked but I'm also pretty sure that all RAW files are opened for **reading only** and so the OS would even reject any writing to them.

**Comment by swsounds** [comment]
parafin

Sorry I missed the comment re the working RAW. Now here it is.

[IMG_2146.CR2.zip](https://github.com/darktable-org/darktable/files/5644271/IMG_2146.CR2.zip)

**Comment by MStraeten** [comment]
can be opened with with darktable on osx without problems. And keeps being ok ;)

**Comment by swsounds** [comment]
@MStraeten Thanks for that info If I have kept exiv2 I'll try and run it as you have from the terminal. So at least I can see if any other problems happen. 

.. and thanks re the 2nd file, which I grabbed off my PC yesterday and imported in a different folder to check that it was ok. And it was.

@TurboGit thanks I have checked info on my RAW (the working one) and got this which says it is Read/write.

![image](https://user-images.githubusercontent.com/75428372/101194553-0db53980-3656-11eb-93a6-de7ac8f72860.png)

I feel that I have learnt a lot this afternoon for which I thank you all. However, I don't know how this RAW got corrupted. It will become clear I'm sure....

**Comment by parafin** [comment]
DamagedFiles is not created by darktable, no. I guess it's created by system disk check utility that gets run automatically after a crash. So file was explicitly marked as corrupted by OS, no wonder darktable couldn't open it.

You talked about crashes, what kind of crashes?

**Comment by TurboGit** [comment]
> @TurboGit thanks I have checked info on my RAW (the working one) and got this which says it is Read/write.

Ok course it is r/w on the OS, I was speaking about the way dt open the files. In the program we open the file read-only and so the created descriptor cannot be used to write on it.

**Comment by swsounds** [comment]
@TurboGit 
Ok thanks for that.

**Comment by swsounds** [comment]
@parafin 
I have a few but it would appear I only sent a couple of them to Apple but here is a copy of one of them (I think). Not on the day the file got corrupted.

If you can understand this please have a look. I did manage to understand one a few weeks ago for Cubase that I had ( I pinned it down to a plugin) but I'm not sure I understand what's going on in these reports.

I was having issues while either drawing a mask or zooming in (not sure which now). I have been having it go slow as well at times. Lots of spinning ball of doom. Also in LT it was not refreshing the thumbnails - getting them as all black or partially scrambled when the images were fine in the DR.

[darktable_2020-11-27-110614_Steves-Mac-mini.crash.zip](https://github.com/darktable-org/darktable/files/5644933/darktable_2020-11-27-110614_Steves-Mac-mini.crash.zip)

**Comment by parafin** [comment]
So it's just applications crashing or did full OS hang happen when you have to do a hard reboot of the computer? Just application crash should affect the filesystem, but if you shut down the computer not normally it might lead to such things. The only thing to check in darktable itself is whether OpenCL is being used - sometimes badly written OS OpenCL drivers lead to whole system crashes.
All-in-all you should close this issue, since it's not a darktable bug, but a problem with your OS.

**Comment by swsounds** [comment]
It was only dt that crashed not the OS hanging as far as I was aware, at least not yesterday. Could you see from the crash report that it wasn't dt?

I'm not using open CL. 

If it's not a dt bug then I'll have to see if it is the tablet that I'm using as that's the only other thing that it can be.

**Comment by ralfbrown** [comment]
> I have been having it go slow as well at times. Lots of spinning ball of doom. Also in LT it was not refreshing the thumbnails - getting them as all black or partially scrambled when the images were fine in the DR.

Such behavior is strongly suggestive of a failing hard drive.  The hard drive itself may spend several seconds trying to get a successful read of a marginal sector; if it returns an error because it couldn't the OS then retries the read a few times (and each time the drive does **its** retries).  Eventually the read indicates success but returns corrupted data, leading to the scrambled display.

**Comment by parafin** [comment]
> It was only dt that crashed not the OS hanging as far as I was aware, at least not yesterday. Could you see from the crash report that it wasn't dt?

I'm saying that dt crash can't cause such corruption.

**Comment by parafin** [comment]
Also data corruption might have happened sometime earlier, when was the last time you opened that file in darkroom?

**Comment by swsounds** [comment]
> Such behavior is strongly suggestive of a failing hard drive. The hard drive itself may spend several seconds trying to get a successful read of a marginal sector; if it returns an error because it couldn't the OS then retries the read a few times (and each time the drive does **its** retries). Eventually the read indicates success but returns corrupted data, leading to the scrambled display.

Thanks for that Ralf. I'll keep an eye on it. I did run verify disk yesterday and it said this partition was ok. I'll keep my eye on it. It is being backed up by time machine so I have a degree of security.

**Comment by swsounds** [comment]
> > It was only dt that crashed not the OS hanging as far as I was aware, at least not yesterday. Could you see from the crash report that it wasn't dt?
> 
> I'm saying that dt crash can't cause such corruption.

Thanks. for all your input.

**Comment by swsounds** [comment]
> Also data corruption might have happened sometime earlier, when was the last time you opened that file in darkroom?

The file was only a day old and I don't think I had a crash that day, slowness maybe but nothing particularly  bothersome. 

I'll follow up with Apple re the DamagedFiles folder and see where that comes from maybe that will give me a clue.

**Comment by swsounds** [comment]
@parafin @TurboGit @chhil @ralfbrown @MStraeten @ptilopteri 

Thanks for all your input yesterday. It was very informative. I will continue to pursue this as a non DT issue unless anything else comes to light in the coming days/weeks. 

All your help is much appreciated. 

Steve

**Comment by chhil** [comment]
Just some info on damaged files folder created by the Mac OS.

https://www.cnet.com/news/should-you-worry-about-a-damaged-files-folder/

**Comment by swsounds** [comment]
> Just some info on damaged files folder created by the Mac OS.
> 
> https://www.cnet.com/news/should-you-worry-about-a-damaged-files-folder/

Thanks. I found something similar on Apple support but the references were quite old. The only odd thing is that the offending folder is not at the root of my HD, but in my Photography folder on my data drive. I am watching it carefully and keeping it all backed up. The system drive is a 5 year old SSD and the data drive is probably the original drive out of the Mac mini that I swapped out to use as the data drive. 

Thanks to all your help I may even get to the bottom of this. 

Thanks again.
